### ISSUE TITLE
Eigenvalue calculation fails with reflectors

### ISSUE THREAD

**Issue by robfairh** [arfc/moltres#114]
When I try to run an eigenvalue calculation, the flux shape I get is clearly wrong.
The geometry is very simple. It is a 1D version of a reactor that has a bottom reflector, fuel, and a top reflector.
I am comparing the Moltres results with a 1D Finite-Difference (FD) solver (and a plot from Serpent from the 3D version of the same reactor).
The predicted eigenvalue is close to what the FD solver predicts, so I don't know if the eigenvalue predicted by Moltres is also wrong or not.

These are the fluxes predicted by Moltres (both fluxes are divided by the group 1 flux max value):
![1D-fuel-reflec-eig1B](https://user-images.githubusercontent.com/43791150/87327762-1caa1c80-c4fa-11ea-90a7-81a90795e999.png)

These are the fluxes predicted by the FD solver (both fluxes are divided by the group 1 flux max value):
![1D-fuel-reflec-eig-FDM](https://user-images.githubusercontent.com/43791150/87327816-3186b000-c4fa-11ea-9b9b-febe7d58079c.png)

These are the fluxes predicted by Serpent in the 3D-reactor (both fluxes are divided by the group 1 flux max value):
![Axial1](https://user-images.githubusercontent.com/43791150/87328763-98589900-c4fb-11ea-97a2-a27a2f67e8d9.png)

This issue can be closed when I figure out (or someone tells me) what is wrong with the Moltres input file.
[This](https://github.com/arfc/moltres/files/4913787/1D-fuel-reflec-eig1.txt) is my Moltres input file, and it can be found in this [link](https://github.com/robfairh/moltres/blob/mhtgr/problems/mhtgr350/1D-neutronics/1D-fuel-reflec-eig1.i).

**Comment by robfairh** [comment]
@smpark7 @joonhon any ideas?

**Comment by robfairh** [comment]
I originally thought that there was a problem with the volumes.
It seemed to me that the calculations were dividing the fluxes by the volumes.
As the volume of the fuel is larger than the volume of the reflectors, then the flux in the active core ends up being flatter than the flux in the reflectors.
However, I modified my mesh so the volumes of the 3 regions are equal, and the result I got is the following:
![1D-fuel-reflec-eig1C](https://user-images.githubusercontent.com/43791150/87329924-66e0cd00-c4fd-11ea-9c95-600778f83b52.png)

**Comment by katyhuff** [comment]
- You say you're making this comparison to a 1D Finite-Difference (FD) solver ... is this something you created?
- Regardless, let's focus on the serpent results as the truth (not the 1d FD solver). It seems like this could be from some geometry approximation. .. . you say the Moltres geometry is 1D? How exactly?
- How did you homogenize the cross sections exactly, for the moltres model?
- I understand that your final plot has 3x more mesh points. Please be certain that you've changed $\Delta z$, not just $n_{points}$. The plot (which has $z_{max}=2400cm$) indicates you've increased the number of points but not $\Delta z$.

**Comment by joonhon** [comment]
@robfairh I'm inclined to think that the 1D FD solver may be the odd one of the three. The reason is that if you're solving a reactor in 1D, you have infinite x and y dimension and that means the reactor has 2 less dimensions to 'leak' the neutrons from and as a result should have a flatter profile as compared to the 3D case. Therefore, the 1D FD result looking like the serpent result suggests that the 1D FD may not be correct. [Also, in your 1D FD, are you solving both gp1 and gp2 simultaneously, in one single matrix?]
Other than the above, differences can also arise if you didn't use the same cross sections and coefficients between your moltres input file and your 1D FD solver...

**Comment by robfairh** [comment]
Thank you guys for your comments. I will try to answer your questions.

@katyhuff:
* You say you're making this comparison to a 1D Finite-Difference (FD) solver ... is this something you created?

Correct. I made the FD solver.

* Regardless, let's focus on the serpent results as the truth (not the 1d FD solver). It seems like this could be from some geometry approximation. .. . you say the Moltres geometry is 1D? How exactly?

My mesh is a line that represents the axial direction of the reactor. This implies, what Alvin described. There is no leakage in the direction perpendicular to z.

* How did you homogenize the cross sections exactly, for the moltres model?

The cross sections were obtained from a 3D model of the reactor. The 3D model produces cross sections of 5 reagions. Inner, outer, bottom, and top reflectors, and homogenized fuel (this comprises the fuel, moderator, and coolant channels).

* I understand that your final plot has 3x more mesh points. Please be certain that you've changed $\Delta z$, not just $n_{points}$. The plot (which has $z_{max}=2400cm$) indicates you've increased the number of points but not $\Delta z$.

That's correct. The number of points changes, but not $$\Delta z$$.

@joonhon:
* The reason is that if you're solving a reactor in 1D, you have infinite x and y dimension and that means the reactor has 2 less dimensions to 'leak' the neutrons from and as a result should have a flatter profile as compared to the 3D case.

I agree with your statement. However, both the Moltres and the FD solver are solving the 1D case. So both of them should see the same effect.

* Also, in your 1D FD, are you solving both gp1 and gp2 simultaneously, in one single matrix?

Yes, they are coupled.

* Other than the above, differences can also arise if you didn't use the same cross sections and coefficients between your moltres input file and your 1D FD solver...

I am using the same set of cross sections for both problems.

Your question gave me an idea @joonhon. I am going to do the following. I will produce the cross sections with serpent for a fuel column (this is a bottom reflector, active core composed of 10 fuel assemblies, and a top reflector) with reflective boundary conditions. Then I will run Moltres input file and the FD solver with those cross sections. The results get closer, and it will be more evident which calculation is wrong.

**Comment by katyhuff** [comment]
> That's correct. The number of points changes, but not $$\Delta z$$.

You should divide $\Delta z$ by 3 if you're multiplying n_points by 3. Otherwise, you're increasing the height of your reactor by 3x. (as seen in increased height in the plot)

**Comment by gridley** [comment]
Judging from the appearance of the flux distribution Moltres is giving you, and the fact that your eigenvalue is correct, I would imagine this is a symptom of not using enough power iterations. Can you include a list of the sequence of eigenvalues computed by the solver? Or maybe you should plot them to see if they have converged. This should be a simple copy-paste from the terminal.

100 power iterations is definitely a bit small for a full-reactor problem. It would be good to calculate the number of fast neutron mean free paths across your reactor is. If that number is large, it will take many power iterations to get a good answer.

**Comment by robfairh** [comment]
I made the model in Serpent of the fuel column (i.e. 10 fuel assemblies stacked on top of each other, bottom, and top reflectors) to generate the cross-sections for Moltres.

The following figure shows the flux calculated with Serpent (both fluxes are divided by the group 1 flux max value). Keff = 1.41933 +/- 0.00006.

![standard-column-detector-AxialB](https://user-images.githubusercontent.com/43791150/88104012-73050400-cb67-11ea-939a-ef33b2e47983.png)

The following figure shows the flux calculated with Moltres (both fluxes are divided by the group 1 flux max value). Keff = 1.407798142. 

![3D-assembly-30-homo-eig_axial](https://user-images.githubusercontent.com/43791150/88104106-9334c300-cb67-11ea-9469-0c0382e8570b.png)

This comparison allowed me to realize that @ajhlee, and my 1D FDM solver was off. I fixed it, its result is anecdotic but here it is anyway:

![1D-assembly-eig-FDM](https://user-images.githubusercontent.com/43791150/88104343-ead32e80-cb67-11ea-8d66-e08286f72edc.png)

Also @gridley you were right about the convergence. The calculation wasn't converged and different tolerances yield slightly different flux shapes.
